Add handling of +1 prioritizing on MS rolls

kofoednielsen opened the pull request:
Hello!

In my guild we have this system where if you win a main spec roll, you get a +1. If you're rolling against someone with a +1 and you don't have a +1 yourself, you win the roll no matter what. If you won 2 main spec rolls, you now have +2 and even +1's will win against your roll no matter what. All the people with the lowest amount of + will roll against each other.

Here are a few examples.
```
Bob MS +1 rolled 99
Alice MS +2 rolled 99
Peter MS +0 rolled 1
Peter wins the roll and is now MS +1
```
```
Bob OS rolled 88
Peter MS +3 rolled 10
Peter wins the roll and is now MS +4
```
```
Bob OS rolled 66
Peter OS rolled 55
Bob wins the roll and no MS + is added
```



My guild is doing this by having a spreadsheet on the side, and then manually calculating who should be in the roll, and manually finding the winner out of those.

This pull request should hopefully be able to replace that spreadsheet, and remove the need of tapping in and out of the game as a master looter.

I've also added the command `/pl` for listing who has MS+ and which items they won to get the +
Also `/pl add <player> <item>` and `/pl rm <player> <item>` for manually adjusting the +1 of players.

![Capture](https://github.com/user-attachments/assets/68c83193-0a4e-4f28-a9c0-2e09f267fac1)
![pl](https://github.com/user-attachments/assets/4d63cf34-abb0-45d2-8ffb-9e50cab15c8c)

sica42 commented:
I suppose you never noticed and ran the tests after you added this. They all fail atm since some of the mock modules being used also needs to be updated to support plus_ones.
If you're familiar with tests it's probably easiest for you to fix yourself. They're all the work of obszczymucha and I've only done minor modifications to them earlier, but I can take a stab at fixing them them if you're unfamiliar with it.

kofoednielsen commented:
I will fix them 👍🏼

coldwhitelight commented:
Hello, first of all thank you for the PR, been wanting to add this myself, but I see you already done it.

I tested the PR, everything seems to work well, including manual changes to +1s with commands and the list.

One thing I actually noticed is when your game crashes, you loose all +1 history. If it happens during raid it causes problems. Is there a possibility to add the +1s to persistent db and maybe a `/pl clear` option? I don't want to add too much work to it, but that would be extra security.

Thank you for your consideration.

kofoednielsen commented:
Hello @coldwhitelight This is already possible, if you Ctrl+Click the dice icon and go the the General tab, you will find "Always keep awards data" at the bottom, as well as a clear button

coldwhitelight commented:
Oh thank you, I didn't even notice that one.

sica42 commented:
The "Always keep awards data" option only prevents the addon from clearing data when joining a new raid or reconnecting after a disconnect. However, it does not protect data during a crash, because the client never gets a chance to save the data to disk in that case. Data is only saved when you log out.
If you want to force a save, you can do a /reload. Of course, to be completely safe, you'd need to reload after each item is awarded...

By the way, did you know you can right-click the roll type in the winners window to change it? So you can toggle off MS to remove an item from +1 as an alternative to using the slash remove command :)

kofoednielsen commented:
Test's no longer fail, however i could/should probably add some to test the new functionality.

kofoednielsen commented on RollFor/src/RollTracker.lua:
```diff
@@ -251,7 +251,7 @@ function M.new( item_on_roll )
     } )
 
     for _, player in ipairs( players or {} ) do
-      add( player.name, player.class, roll_type, player.plus_ones )
```
Actually the tests helped me discover this mistake

coldwhitelight commented:
> The "Always keep awards data" option only prevents the addon from clearing data when joining a new raid or reconnecting after a disconnect. However, it does not protect data during a crash, because the client never gets a chance to save the data to disk in that case. Data is only saved when you log out.
If you want to force a save, you can do a /reload. Of course, to be completely safe, you'd need to reload after each item is awarded...

That is actually a little bit unfortunate. I thought is updates the persistent db every time you award an item. Oh well, there's room for upgrade then, thanks for the explanation.

> By the way, did you know you can right-click the roll type in the winners window to change it? So you can toggle off MS to remove an item from +1 as an alternative to using the slash remove command :)

That actually doesn't work for me with this PR (or I'm just dumb). No right click on any window/text/part switches the roll type.

sica42 commented:
> That is actually a little bit unfortunate. I thought is updates the persistent db every time you award an item. Oh well, there's room for upgrade then, thanks for the explanation.
Agree, but I don't think there's any easy solution for it. It does update the persistent db, but there is no way for an addon to control when that gets written to disk. It only happens on logout and /reload.
I have been thinking about maybe sending the winner info to other clients so that you can restore it on a crash, but that's a lot of work for something that doesn't happen very often.
> 
> > By the way, did you know you can right-click the roll type in the winners window to change it? So you can toggle off MS to remove an item from +1 as an alternative to using the slash remove command :)
> 
> That actually doesn't work for me with this PR (or I'm just dumb). No right click on any window/text/part switches the roll type.
Hmm, that's strange. The winner popup should have right click on the headers (to change filters) and on the type for each item. Just to clarify, you have to right click the OS/OS/MS text for each item. Do none of those work for you?

sica42 commented:
Oh, one thing I thought of while testing this. The confirmation dialog if you wan to add it to +1 is nice, but maybe there should be an option to toggle it off and just add everything to +1 if that's what you want?

kofoednielsen commented:
> Oh, one thing I thought of while testing this. The confirmation dialog if you wan to add it to +1 is nice, but maybe there should be an option to toggle it off and just add everything to +1 if that's what you want?

100% agree, I'm working on making that optional

sica42 commented:
Found a bug:
`ERROR: Interface\AddOns\RollFor\src\RollingPopup.lua:185: attempt to compare number with nil`
Need a nil check on v.plus_ones there.

Btw, when you feel you're done with all the features you want can you update version to 4.8 and squash your commits? Then I'd be happy to merge it.

kofoednielsen commented:
Hello,

I made the +1 prompt optional with the setting `Always prompt for +1's`. With this disabled, any MS award will give a +1. I found it annoying that tooltips are not supported with options that hide when another optoin is disabled.

I also squashed everything to a single commit and updated the version number

I tried to retproduce your bug with a lot of different rolls: MS, OS, MS ties, OS ties, SR and SR ties but i could not reproduce it. Do you remember what you did? I don't want to just add the `nil` check, because by my design `plus_ones` should never be `nil`. It's gets set to 0 when +1 handling is disabled. 
https://github.com/sica42/roll-for-vanilla/blob/c8ef435abd7cd74d1ab7884c1f90aa24b447d4f0/RollFor/src/RollingStrategyFactory.lua#L59-L65

I did find a single case where `plus_ones` was not passed to a function expecting it, that I fixed before performing my testing, so it could be that this was the problem. The problem was on this line, but is already fixed.
https://github.com/sica42/roll-for-vanilla/blob/c8ef435abd7cd74d1ab7884c1f90aa24b447d4f0/RollFor/src/TieRollingLogic.lua#L146

I will do some more testing but otherwise i think this is ready.

kofoednielsen commented:
Tested it out on like 8 raids, no issues.

sica42 commented on RollFor/RollFor-BCC.toc:
```diff
@@ -1,7 +1,7 @@
 ## Title: RollFor
 ## Author: Obszczymucha
-## Version: 4.7.12
+## Version: 4.8.t :wq0
```
Some weird characters at the end here

sica42 commented on RollFor/src/Client.lua:
```diff
@@ -307,7 +308,7 @@ function M.new( ace_timer, player_info, rolling_popup, config )
         end
 
-        roll_tracker.add( data.player_name, data.player_class, data.roll_type, data.roll )
+        roll_tracker.add( data.player_name, data.player_class, data.roll_type, data.roll, data.plus_ones )
```
This is probably where the bug I experienced originated from. I was using this version, while LM was using an older version, so data.plus_ones would defiantly be nil.
`data.plus_ones or false` should probably fix it.

kofoednielsen replied:
Didn't realise that this feature existed:
> Show roll popup for all group/raid members who have the addon installed when loot master starts a roll.
Enable with /rf config client show-roll Eligible or Always
/rf config client to view additional client options

I've added a `or 0` because this is `plus_ones` which is `the amount of plus_ones the rolling player has`. In other places there is a variable called `plus_one` which is `should this award count as a +1`.

sica42 commented:
Hey.
I've been on vacation so sorry for not replying.

See my 2 comments above. Get those fixed and I'll merge it

kofoednielsen commented:
Done :)
